src/main/scala/nl/biopet/tools/digenicsearch/DigenicSearch.scala

```diff
@@ -96,36 +120,72 @@ object DigenicSearch extends ToolCommand[Args] {
       .toDS()
   }
 
+  def outputFamilyGenes(outputDir: File): File =
+    new File(outputDir, "familyGenes.tsv")
+  def outputAggregation(outputDir: File): File =
+    new File(outputDir, "aggregation")
   def outputPairs(outputDir: File): File = new File(outputDir, "pairs")
   def outputVariants(outputDir: File): File = new File(outputDir, "variants")
 
-  def aggregate(cmdArgs: Args,
-                variants: Dataset[Variant],
-                dict: SAMSequenceDictionary)(
-      implicit sparkSession: SparkSession): Unit = {
+  def createAggregateRegions(aggregationFile: Option[File],
+                             dict: SAMSequenceDictionary)(
+      implicit sparkSession: SparkSession): Option[Dataset[Region]] = {
     import sparkSession.implicits._
-    cmdArgs.aggregation.foreach { file =>
-      val bedrecords = sparkSession.sparkContext
+    aggregationFile.map { file =>
+      sparkSession.sparkContext
         .parallelize(Region.fromBedFile(file, dict))
         .toDS()
-
-      val j =
-        bedrecords
-          .joinWith(variants,
-                    bedrecords("contig") === variants("contig") && bedrecords(
-                      "start") <= variants("pos") && bedrecords("end") >= variants(
-                      "pos"))
-          .rdd
-          .map { case (r, v) => r.name -> v }
-          .groupByKey()
-          .map { case (g, l) => GeneCounts(g, l.size) }
-          .toDS()
-
-      val outputFile = new File(cmdArgs.outputDir, "aggregation")
-      j.write.csv(outputFile.getAbsolutePath)
+        .cache()
     }
   }
 
+  def createAggregateFamilies(aggregateRegion: Dataset[Region],
+                              variants: Dataset[Variant],
+                              broadcasts: Broadcast[Broadcasts])(
+      implicit sparkSession: SparkSession): Dataset[GeneFamilyCounts] = {
+    import sparkSession.implicits._
+    aggregateRegion
+      .joinWith(
```

Comment:
A lot of .method.method.method.method.method calls here. Can you put in comments that explain what each step is doing so it can easily be debugged later?

Reply:
Not all of them because a lot of them are spark specific. Still added a comment.